tiledb/common/memory_tracker.cc

```diff
@@ -0,0 +1,319 @@
+  total_counter_.fetch_add(bytes, std::memory_order_relaxed);
+  type_counter_.fetch_add(bytes, std::memory_order_relaxed);
+  return upstream_->allocate(bytes, alignment);
+}
+
+void MemoryTrackerResource::do_deallocate(
+    void* ptr, size_t bytes, size_t alignment) {
+  upstream_->deallocate(ptr, bytes, alignment);
+  type_counter_.fetch_sub(bytes, std::memory_order_relaxed);
+  total_counter_.fetch_sub(bytes, std::memory_order_relaxed);
+}
+
+bool MemoryTrackerResource::do_is_equal(
+    const tdb::pmr::memory_resource& other) const noexcept {
+  return this == &other;
+}
+
+MemoryTracker::~MemoryTracker() {
+  assert(
+      total_counter_.fetch_add(0) == 0 &&
+      "MemoryTracker destructed with outstanding allocations.");
+}
+
+tdb::pmr::memory_resource* MemoryTracker::get_resource(MemoryType type) {
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  // If we've already created an instance for this type, return it.
+  auto iter = resources_.find(type);
+  if (iter != resources_.end()) {
+    return iter->second.get();
+  }
+
+  // Add a new counter if it doesn't exist.
+  if (counters_.find(type) == counters_.end()) {
+    counters_.emplace(type, 0);
+  } else {
+    // There's no outstanding memory resource for this type, so it must be zero.
+    assert(counters_[type] == 0 && "Invalid memory tracking state.");
+  }
+
+  // Create and track a shared_ptr to the new memory resource.
+  auto ret = make_shared<MemoryTrackerResource>(
+      HERE(), upstream_, total_counter_, counters_[type]);
+  resources_.emplace(type, ret);
+
+  // Return the raw memory resource pointer for use by pmr containers.
+  return ret.get();
+}
+
+std::tuple<uint64_t, std::unordered_map<MemoryType, uint64_t>>
+MemoryTracker::get_counts() {
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  auto total = total_counter_.fetch_add(0, std::memory_order_relaxed);
+  std::unordered_map<MemoryType, uint64_t> by_type;
+  std::vector<MemoryType> to_del;
+  for (auto& [mem_type, resource] : resources_) {
+    by_type[mem_type] = resource->get_count();
+  }
+
+  return {total, by_type};
+}
+
+uint64_t MemoryTracker::generate_id() {
+  static std::atomic<uint64_t> curr_id{0};
+  return curr_id.fetch_add(1);
+}
+
+shared_ptr<MemoryTracker> MemoryTrackerManager::create_tracker() {
+  class MemoryTrackerCreator : public MemoryTracker {
+   public:
+    MemoryTrackerCreator()
+        : MemoryTracker() {
+    }
+  };
+
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  // Delete any expired weak_ptr instances
+  size_t idx = 0;
+  while (idx < trackers_.size()) {
+    if (trackers_[idx].expired()) {
+      trackers_.erase(trackers_.begin() + idx);
+    } else {
+      idx++;
+    }
+  }
+
+  // Create a new tracker
+  auto ret = make_shared<MemoryTrackerCreator>(HERE());
+  trackers_.emplace(trackers_.begin(), ret);
+
+  return ret;
+}
+
+std::string MemoryTrackerManager::to_json() {
+  std::lock_guard<std::mutex> lg(mutex_);
+  nlohmann::json rv;
+
+  size_t idx = 0;
```

Comment:
Switch to a for loop? Also add a comment:
// Iterate through all the trackers.

Reply:
This was originally a for-loop but that caused a subtle but terrible bug while iterating over a vector and modifying it simultaneously. The underlying issue here is that we're discarding invalid `weak_ref`'s while iterating. However, given that you suggested the change means this is a good place to add an explanatory comment on why its done with a while loop instead.

The fix for the bug is in this commit:

https://github.com/TileDB-Inc/TileDB/pull/4683/commits/f5a037883fc89767f23cc79a532da8b5d042d7d1

Reply:
I see, current loop is fine.

Reply:
Added a comment for future us.